tools/go/gozip/main.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"archive/zip"
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+)
+
+var (
+	fbaseDir   string
+	fdir       string
+	finputFile string
+	foutput    string
+	fquiet     bool
+)
+
+func init() {
+	flag.StringVar(&fbaseDir, "base-dir", "", "Base dir for the zip archive")
+	flag.StringVar(&fdir, "dir", "", "dir to zip")
+	flag.StringVar(&finputFile, "input-file", "", "a file containing a list of files to archive")
+	flag.StringVar(&foutput, "output", "", "Output file name")
+	flag.BoolVar(&fquiet, "quiet", false, "no output")
+}
+
+func validateFlags() {
+	if fdir == "" && finputFile == "" && fbaseDir == "" && foutput == "" {
+		flag.Usage()
+		os.Exit(0)
+	}
+
+	if foutput == "" {
+		flag.Usage()
+		panic("output file name is required")
+	}
+
+	if fdir == "" && finputFile == "" {
+		flag.Usage()
+		panic("file and path can't both be empty")
+	}
+
+	if fdir != "" && finputFile != "" {
+		flag.Usage()
+		panic("file and path are mutually exclusive")
+	}
+
+	if fbaseDir == "" && fdir != "" {
+		fbaseDir = fdir
+	}
+
+	if !strings.HasSuffix(fbaseDir, "/") {
+		fbaseDir = fbaseDir + "/"
+	}
+}
+
+func zipDir(filename, dir string) error {
+	newZipFile, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer newZipFile.Close()
+
+	zipWriter := zip.NewWriter(newZipFile)
+	defer zipWriter.Close()
+
+	if err = addFiles(zipWriter, dir, dir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func zipFiles(filename string, files []string, baseInZip string) error {
+	newZipFile, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer newZipFile.Close()
+
+	zipWriter := zip.NewWriter(newZipFile)
+	defer zipWriter.Close()
+
+	for _, file := range files {
+		if err = addFileToZip(zipWriter, file, baseInZip); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addFileToZip(zipWriter *zip.Writer, filePath, baseInZip string) error {
+	fileToZip, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer fileToZip.Close()
+
+	info, err := fileToZip.Stat()
+	if err != nil {
+		return err
+	}
+
+	header, err := zip.FileInfoHeader(info)
+	if err != nil {
+		return err
+	}
+
+	baseLength := len(baseInZip)
+	if !strings.HasSuffix(baseInZip, "/") {
```

Comment:
Would `go` always use `/` separator regardless of the OS? I am guessing yes, because of the use of `path`. But just wanted to confirm. :)

Reply:
good point. I'll fix it

Reply:
good point. I'll fix it